File: spring_server/src/main/java/com/semothon/spring_server/crawling/entity/Crawling.java
package com.semothon.spring_server.crawling.entity;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.semothon.spring_server.chat.entity.ChatRoom;
import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor
@Builder
@ToString(of = {"crawlingId", "title", "url", "deadlinedAt", "crawledAt"})
@Table(name = "crawling",
        indexes = {
        },
        uniqueConstraints = {
                @UniqueConstraint(columnNames = {"url"})
        }
)
public class Crawling {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long crawlingId;

    @Column(nullable = false, length = 255)
    private String title;

    @Column(nullable = false, length = 500)
    private String url;

    private String imageUrl;

    @Column(nullable = false, columnDefinition = "TEXT")
    private String description;

    @Column(columnDefinition = "TIMESTAMP")
    private LocalDateTime deadlinedAt;

    @Column(columnDefinition = "TIMESTAMP")
    private LocalDateTime crawledAt;

    @JsonIgnore
    @Builder.Default
    @OneToMany(mappedBy = "crawling", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<UserCrawlingRecommendation> userCrawlingRecommendations = new ArrayList<>();

    @JsonIgnore
    @Builder.Default
    @OneToMany(mappedBy = "crawling", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<ChatRoom> chatRooms = new ArrayList<>();

    @JsonIgnore
    @Builder.Default
    @OneToMany(mappedBy = "crawling", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<CrawlingInterest> crawlingInterests = new ArrayList<>();

    protected void addUserCrawlingRecommendation(UserCrawlingRecommendation userCrawlingRecommendation){
        this.userCrawlingRecommendations.add(userCrawlingRecommendation);
    }

    protected void addCrawlingInterest(CrawlingInterest crawlingInterest){
        this.crawlingInterests.add(crawlingInterest);
    }

    public void addChatRoom(ChatRoom chatRoom){
        this.chatRooms.add(chatRoom);
    }
}

File: fastapi_server/app/models/crawling.py
from sqlalchemy import Column, String, Integer, Text, DateTime
from sqlalchemy.orm import relationship
from database import Base
from custom_descriptable import CrawlingDescriptable

class Crawling(Base):
    __tablename__ = "crawling"

    crawling_id = Column(Integer, primary_key=True, autoincrement=True)
    title = Column(String(255), nullable=False)
    url = Column(String(500), nullable=False)
    image_url = Column(String, nullable=True)
    description = Column(Text, nullable=False)
    deadlined_at = Column(DateTime, nullable=True)
    crawled_at = Column(DateTime, nullable=True)

    crawling_interests = relationship("CrawlingInterest", cascade="all, delete-orphan")

    user_crawling_recommendations = relationship(
        "UserCrawlingRecommendation",
        cascade="all, delete-orphan"
    )

def crawling_to_descriptable(crawling: Crawling) -> dict:
    return CrawlingDescriptable(crawling.description, crawling.crawling_id)
